lib/plug/conn.ex

```diff
@@ -934,6 +934,50 @@ defmodule Plug.Conn do
   end
 
   @doc """
+  Push a resource to the client. Server pushes must happen prior to a response
+  being sent. If a server push is attempted after a response is sent then a
+  `Plug.Conn.AlreadySentError` will be raised.
+
+  If the adapter does not support server push then this is a noop.
+  """
+  @spec push(t, String.t, Keyword.t) :: t
+  def push(conn, path, headers \\ []) do
+    adapter_push(conn, path, headers)
```

Comment:
I think we should get the req back from the adapter and put it back. Cowboy doesn't seem to mutate conn but maybe another adapter would (such as the test one as proposed above!)